### PR TITLE
Feature/work-buddy

### PR DESCRIPTION
**Pull request by TheaAleksovDev**
practicing useRef, useReducer, memo, useCallback, useMemo, useContext

### CODE REVIEW

File: src/components/BuddiesContext.tsx
Comment (valentin-aleksandrov) [inline src/components/BuddiesContext.tsx:1]: Do you think we need the `setData` function? Can we directly work with `setContextData` function?
image

File: src/components/BuddiesContext.tsx [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import { Buddy } from "./WorkBuddies/BuddyInterface";
+
+interface BuddiesContextType {
+  contextData: Buddy[] | undefined;
+  setData: (data: Buddy[]) => void;
+}
+
+export const BuddiesContext = createContext<BuddiesContextType>({
+  contextData: [],
+  setData: (data: Buddy[]) => {},
```
Comment (valentin-aleksandrov) [inline src/components/BuddiesContext.tsx]: This can be simplified like so:
 ```suggestion
  setData: () => {},
```

File: src/components/BuddiesContext.tsx [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import React, { createContext, useEffect, useState } from "react";
```
Comment (valentin-aleksandrov) [inline src/components/BuddiesContext.tsx]: Since React 18, this is no longer required to be imported:
```suggestion
import { createContext, useEffect, useState } from "react";
```

File: src/components/Tasks/Tasks.tsx [hunk 3]
```diff
@@ -70,19 +69,40 @@ const Tasks = () => {
...
-      <Task
-        key={item.task}
-        {...item}
-        markCompleted={markCompleted}
-        deleteTask={deleteTask}
-      />
-    );
-  });
+   //useMemo
+  const displayTasks = useMemo(()=>{
```
Comment (valentin-aleksandrov) [inline src/components/Tasks/Tasks.tsx]: This is something that I have missed in the last PR. 
We use verbs for functions.
Since this isn't a function, it should be named differently.
One way would be:
```suggestion
  const tasksToDisplay = useMemo(()=>{
```

File: src/components/Tasks/Tasks.tsx [hunk 4]
```diff
@@ -70,19 +69,40 @@ const Tasks = () => {
...
+          {...item}
+          markCompleted={markCompleted}
+          deleteTask={deleteTask}
+        />
+      );
+    });
+  },[filteredTasks, tasks])
+
+  //useMemo
+  const displayContext = useMemo(()=>{
```
Comment (valentin-aleksandrov) [inline src/components/Tasks/Tasks.tsx]:
```suggestion
  const buddies = useMemo(()=>{
```

File: src/components/Tasks/Tasks.tsx [hunk 5]
```diff
@@ -70,19 +69,40 @@ const Tasks = () => {
...
+        <div className="buddy-smaller" key={index}>
+          <img src={buddy.url} alt="buddy-image" />
+        </div>
+      );
+    });
+  },[contextData])
 
   return (
     <div className="tasks">
+      <div className="buddies-smaller">{contextData && displayContext}</div>
```
Comment (valentin-aleksandrov) [inline src/components/Tasks/Tasks.tsx]: Do you see a problem if simplify it like so:
```suggestion
      <div className="buddies-smaller">{displayContext}</div>
```

File: src/components/WorkBuddies/MyWorkBuddy.tsx [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+      return 1;
+  }
+};
+
+const MyWorkBuddy = memo(({ myBuddy }: propsInterface) => {
+  const { contextData } = useContext(BuddiesContext);
+
+  //useReducer
+  const [currentBuddyIndex, dispatch] = useReducer(reducer, 0);
+  const [buddy, setBuddy] = useState<Buddy>();
```
Comment (valentin-aleksandrov) [inline src/components/WorkBuddies/MyWorkBuddy.tsx]: If I remove this line and the bellow useEffect, would anything change in the 
application?
image

The reason I'm asking is because I don't see the `buddy` variable used anywhere.
It's almost as if you are setting a state and not using it.

File: src/components/WorkBuddies/MyWorkBuddy.tsx [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+  //but I included both concepts just for practice (memo and useContext)
+
+  useEffect(() => {
+    if (myBuddy) {
+      setBuddy(myBuddy);
+    }
+  }, [myBuddy]);
+
+  useEffect(() => {
+    if (!contextData) return;
```
Comment (valentin-aleksandrov) [inline src/components/WorkBuddies/MyWorkBuddy.tsx]: This is a valid syntax.
I would encourage you to write like this:
```suggestion
    if (!contextData) {
        return;
    }
```
In my experience Developers are finding this more readable than a one liner.

File: src/components/WorkBuddies/MyWorkBuddy.tsx [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+
+const MyWorkBuddy = memo(({ myBuddy }: propsInterface) => {
+  const { contextData } = useContext(BuddiesContext);
+
+  //useReducer
+  const [currentBuddyIndex, dispatch] = useReducer(reducer, 0);
+  const [buddy, setBuddy] = useState<Buddy>();
+  const [myBuddyIndex, setMyBuddyIndex] = useState<number>();
+
+  //I am aware that I can get "buddy" from contextData and that I'm not actually using it,
```
Comment (valentin-aleksandrov) [inline src/components/WorkBuddies/MyWorkBuddy.tsx]: Is this related to my question on line **45**?

File: src/components/WorkBuddies/WorkBuddies.tsx [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+      });
+
+      setBuddies(todaysBuddies);
+    } catch (error) {
+      console.log(error);
+    }
+  }, [setBuddies]);
+
+  //useMemo
+  const displayBuddies = useMemo(() => {
```
Comment (valentin-aleksandrov) [inline src/components/WorkBuddies/WorkBuddies.tsx]:
```suggestion
  const buddiesToDisplay = useMemo(() => {
```

File: src/components/WorkBuddies/WorkBuddies.tsx [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+        name: nameRef.current!.value,
+        isMine: true,
+      }));
+    }
+
+    if (myBuddy === undefined) {
+      return;
+    }
+
+    const updatedBuddies = buddies;
```
Comment (valentin-aleksandrov) [inline src/components/WorkBuddies/WorkBuddies.tsx]: This doesn't work the way you are expecting.
JS works with references.
If you want to create a new copy of the array so that you want to mutate it, you should do something like so:
```suggestion
    const updatedBuddies = [...buddies];
```

File: src/components/WorkBuddies/WorkBuddies.tsx [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+    updatedBuddies.forEach((buddy) => {
```
Comment (valentin-aleksandrov) [inline src/components/WorkBuddies/WorkBuddies.tsx]: Why not use **.map**?
.map will automatically create a new array so that you're not mutating the original.

File: src/components/WorkBuddies/WorkBuddies.tsx [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+        </div>
+      );
+    });
+  }, [buddies, myBuddy]);
+
+  const onChooseOne = () => {
+    if (nameRef.current && nameRef.current.value) {
+      setMyBuddy((prev) => ({
+        ...(prev as Buddy),
+        name: nameRef.current!.value,
```
Comment (valentin-aleksandrov) [inline src/components/WorkBuddies/WorkBuddies.tsx]: What you have created as example of `useRef` is something that can be avoided by using state. I would encourage to rewrite this using a state and think about
something different for using `useRef`.

For example useRef gives you access to the actual DOM element.
Why not implement something that can't be implemented via React?
Some ideas:
- scroll element into view
- focus element
- submit a form programmatically 
- others.

What I'm saying is that taking the value of an input is not a good scenario for useRef, because we already have an established way of how to do that with React's onChange event on inputs and React's useState hook.
We need something unique.
React doesn't recommend using `useRef`, unless there is no other way.

File: src/components/WorkBuddies/MyWorkBuddy.tsx [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+      if (action.myBuddyIndex) {
+        return action.myBuddyIndex;
+      } else return 0;
+
+    default:
+      return 1;
+  }
+};
+
+const MyWorkBuddy = memo(({ myBuddy }: propsInterface) => {
```
Comment (valentin-aleksandrov) [inline src/components/WorkBuddies/MyWorkBuddy.tsx]: image

You may need to take the individual props of the `myBuddy` object

Objects can have different references even if they have the same properties.
The consequence of that is **memo** becomes useless.

If you want `memo` to actually work, you can expect primitives:
    - name: string;
    - url: string;
    - isMine?: boolean;

File: src/components/WorkBuddies/MyWorkBuddy.tsx [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+    if (!contextData) return;
+    contextData.forEach((element, index) => {
+      if (element.isMine) {
+        setMyBuddyIndex(index);
+      }
+    });
+  }, [contextData]);
+
+  return (
+    <Fragment>
```
Comment (valentin-aleksandrov) [inline src/components/WorkBuddies/MyWorkBuddy.tsx]:
```suggestion
    <>
```

You would need to use `Fragment` if you were using it in a .map method and you wanted to pass a `key` prop to it.

File: src/components/WorkBuddies/WorkBuddies.tsx [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+  useEffect(() => {
+    if (isBuddyChosen) {
+      setData(buddies);
+    }
+  }, [isBuddyChosen, setData, buddies]);
+
+  return (
+    <div className="work-buddies">
+      {!buddies.length && !isBuddyChosen && (
+        <Fragment>
```
Comment (valentin-aleksandrov) [inline src/components/WorkBuddies/WorkBuddies.tsx]:
```suggestion
        <>
```

File: src/components/WorkBuddies/WorkBuddies.tsx [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+          <button className="get-options green" onClick={fetchBuddies}>
+            get options
+          </button>
+        </Fragment>
+      )}
+      {buddies && !isBuddyChosen && (
+        <div className="buddies-container">{displayBuddies}</div>
+      )}
+      {buddies.length > 1 && !isBuddyChosen && (
+        <Fragment>
```
Comment (valentin-aleksandrov) [inline src/components/WorkBuddies/WorkBuddies.tsx]:
```suggestion
        <>
```

File: src/components/WorkBuddies/MyWorkBuddy.tsx (hunk 13 above)
Comment (valentin-aleksandrov) [reply src/components/WorkBuddies/MyWorkBuddy.tsx]: If you need to pass objects, React allows you to pass a custom comparison function:
- image

File: src/components/WorkBuddies/MyWorkBuddy.tsx [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+
+const reducer = (
+  index: number,
+  action: { type: string; contextData?: Buddy[]; myBuddyIndex?: number }
+) => {
+  switch (action.type) {
+    case "next":
+      if (action.contextData) {
+        if (index + 1 < action.contextData.length) return index + 1;
+        else return 0;
```
Comment (valentin-aleksandrov) [inline src/components/WorkBuddies/MyWorkBuddy.tsx]: Please, specify { }

File: src/components/WorkBuddies/MyWorkBuddy.tsx [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+      } else return 0;
+
+    case "prev":
+      if (action.contextData) {
+        if (index - 1 > -1) return index - 1;
+        else return action.contextData.length - 1;
```
Comment (valentin-aleksandrov) [inline src/components/WorkBuddies/MyWorkBuddy.tsx]: Can all of these be simplified with the ternary operator? 
```suggestion
        return (index - 1 > -1) ? index - 1 : action.contextData.length - 1;
```

File: src/components/BuddiesContext.tsx
Comment (TheaAleksovDev) [reply src/components/BuddiesContext.tsx:1]: I would never have thought of that. I think I've never done it that simply. My other approach would probably be:
     `<BuddiesContext.Provider value={{ contextData, setData: (data)=>{setContextData(data)} }}>`

File: src/components/WorkBuddies/WorkBuddies.tsx (hunk 9 above)
Comment (TheaAleksovDev) [reply src/components/WorkBuddies/WorkBuddies.tsx]: Is this a common naming convention when using useMemo?

File: src/components/WorkBuddies/MyWorkBuddy.tsx (hunk 13 above)
Comment (TheaAleksovDev) [reply src/components/WorkBuddies/MyWorkBuddy.tsx]: Got it. I will pass buddy's name, then conditionally change color of the name if it's mine, and that way I'll actually use it in this component and resolve comments that you made about it.

File: src/components/Tasks/Tasks.tsx (hunk 3 above)
Comment (TheaAleksovDev) [reply src/components/Tasks/Tasks.tsx]: Oh okay. I asked you about that naming convetion in WorkBuddies.tsx , but now I get it

File: src/components/Tasks/Tasks.tsx (hunk 5 above)
Comment (TheaAleksovDev) [reply src/components/Tasks/Tasks.tsx]: I'm just making sure they'll be displayed only if contextData exists, but I guess that's fine too.
Sometimes I'm getting errors that it may not exist, and I started doing that as a prevention.

File: src/components/BuddiesContext.tsx
Comment (valentin-aleksandrov) [reply src/components/BuddiesContext.tsx:1]: What I'm suggesting is called Point-free style programming.
You can read more about it here:
- https://medium.com/dailyjs/functional-js-7-point-free-style-b21a1416ac6a

File: src/components/WorkBuddies/WorkBuddies.tsx (hunk 9 above)
Comment (valentin-aleksandrov) [reply src/components/WorkBuddies/WorkBuddies.tsx]: No. The idea is that this is something you want to display on the screen.
The best name would be just `buddies`. Since, that is already taken, you can use something like `buddiesToDisplay`

File: src/components/WorkBuddies/MyWorkBuddy.tsx (hunk 8 above)
Comment (TheaAleksovDev) [reply src/components/WorkBuddies/MyWorkBuddy.tsx]: Yes, but I changed it.

File: src/components/WorkBuddies/MyWorkBuddy.tsx (hunk 14 above)
Comment (TheaAleksovDev) [reply src/components/WorkBuddies/MyWorkBuddy.tsx]: I'm not sure that I'm understanding this.
Should I use Fragment in every display .map and add key to it?
Or should I just add keys to Fragments that I'm using in component's main return?

File: src/components/WorkBuddies/MyWorkBuddy.tsx (hunk 6 above)
Comment (TheaAleksovDev) [reply src/components/WorkBuddies/MyWorkBuddy.tsx]: It was just for practice of using useMemo(), but after reading other suggestions I changed it to name:string instead of myBuddy, and now I'm using it

File: src/components/WorkBuddies/MyWorkBuddy.tsx (hunk 14 above)
Comment (valentin-aleksandrov) [reply src/components/WorkBuddies/MyWorkBuddy.tsx]: What I'm suggestion is to always use the following form of a Fragment if it is needed to use a fragment.
```jsx
<>
  <h2>Title</h2>
  <p>Some Text</p>
</>
```

Comment (valentin-aleksandrov) [reply src/components/WorkBuddies/MyWorkBuddy.tsx]: The only scenario, where you need to use the **\<Fragment>** syntax over the **<>** one is when you are passing a key prop.
```jsx
   {
       [1, 2, 3].map(element => (<Fragment key={element}>...</Fragment>))
   }
```

Comment (TheaAleksovDev) [reply src/components/WorkBuddies/MyWorkBuddy.tsx]: Ooh, got it. Thanks.

File: src/components/WorkBuddies/WorkBuddies.tsx [hunk 19]
```diff
@@ -103,34 +110,39 @@ const WorkBuddies = () => {
     if (isBuddyChosen) {
       setData(buddies);
     }
+
+    if (buddies && !isBuddyChosen) {
+      nameRef.current?.focus();
```
Comment (valentin-aleksandrov) [inline src/components/WorkBuddies/WorkBuddies.tsx:111]: Awesome!

File: src/components/WorkBuddies/WorkBuddies.tsx [hunk 20]
```diff
@@ -85,13 +90,15 @@ const WorkBuddies = () => {
...
-    updatedBuddies.forEach((buddy) => {
+    const updatedBuddies = buddies.map((buddy) => {
       if (buddy.url === myBuddy.url) {
-        if (nameRef.current?.value) {
-          buddy.name = nameRef.current!.value;
-        }
-        buddy.isMine = true;
+        return {
+          ...buddy,
+          name: nameRef.current?.value ? nameRef.current?.value : buddy.name,
```
Comment (valentin-aleksandrov) [inline src/components/WorkBuddies/WorkBuddies.tsx]: I'm expecting here the `name` variable to be used instead of the nameRef.current?.value

File: src/components/WorkBuddies/WorkBuddies.tsx [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+          }}
+        >
+          <img src={buddy.url} alt="buddy-image" />
+        </div>
+      );
+    });
+  }, [buddies, myBuddy]);
+
+  const onChooseOne = () => {
+    if (nameRef.current && nameRef.current.value) {
```
Comment (valentin-aleksandrov) [inline src/components/WorkBuddies/WorkBuddies.tsx]: Since you have a state for that, I would expect to use it, instead of the ref.

File: src/components/WorkBuddies/WorkBuddies.tsx [hunk 22]
```diff
@@ -0,0 +1,150 @@
...
+      setMyBuddy((prev) => ({
+        ...(prev as Buddy),
+        name: nameRef.current!.value,
```
Comment (valentin-aleksandrov) [inline src/components/WorkBuddies/WorkBuddies.tsx]: Same as my above comment.

File: src/components/WorkBuddies/MyWorkBuddy.tsx [hunk 23]
```diff
@@ -0,0 +1,122 @@
...
+  action: { type: string; contextData?: Buddy[]; myBuddyIndex?: number }
+) => {
+  switch (action.type) {
+    case "next":
+      if (action.contextData) {
+        if (index + 1 < action.contextData.length) {
+          return index + 1;
+        } else {
+          return 0;
+        }
```
Comment (valentin-aleksandrov) [inline src/components/WorkBuddies/MyWorkBuddy.tsx]:
```suggestion
        return (index + 1 < action.contextData.length) ? index + 1 : 0
```
When addressing PR suggestions, think about what other places you can apply the received suggestions.

File: src/components/WorkBuddies/MyWorkBuddy.tsx [hunk 24]
```diff
@@ -0,0 +1,122 @@
...
+
+    case "prev":
+      if (action.contextData) {
+        return index - 1 > -1 ? index - 1 : action.contextData.length - 1;
+      } else return 0;
+
+    case "my-buddy":
+      if (action.myBuddyIndex) {
+        return action.myBuddyIndex;
+      } else return 0;
```
Comment (valentin-aleksandrov) [inline src/components/WorkBuddies/MyWorkBuddy.tsx]:
```suggestion
      return (action.myBuddyIndex) ? action.myBuddyIndex : 0;
```

File: src/components/WorkBuddies/MyWorkBuddy.tsx [hunk 25]
```diff
@@ -0,0 +1,122 @@
...
+  //useReducer
+  const [currentBuddyIndex, dispatch] = useReducer(reducer, 0);
+  const [BuddysName, setBuddysName] = useState<string>();
+  const [myBuddyIndex, setMyBuddyIndex] = useState<number>();
+
+  useEffect(() => {
+    if (name) {
+      setBuddysName(name);
+    }
+  }, [name]);
```
Comment (valentin-aleksandrov) [inline src/components/WorkBuddies/MyWorkBuddy.tsx]: Can we simplify this like so:
image

File: src/components/Tasks/Tasks.tsx (hunk 5 above)
Comment (valentin-aleksandrov) [reply src/components/Tasks/Tasks.tsx]: We want to avoid unneeded preventions, because later we would need to write down unneeded tests and it's making the code more complicated than it needs to be.

File: src/components/WorkBuddies/MyWorkBuddy.tsx (hunk 25 above)
Comment (TheaAleksovDev) [reply src/components/WorkBuddies/MyWorkBuddy.tsx:50]: yes, absolutely..

File: src/components/WorkBuddies/WorkBuddies.tsx [hunk 26]
```diff
@@ -0,0 +1,150 @@
...
+import { Buddy } from "./BuddyInterface";
+import { BuddiesContext } from "../BuddiesContext";
+import MyWorkBuddy from "./MyWorkBuddy";
+
+const WorkBuddies = () => {
+  const [buddies, setBuddies] = useState<Buddy[]>([]);
+  const [myBuddy, setMyBuddy] = useState<Buddy>();
+  const [isBuddyChosen, setIsBuddyChosen] = useState<boolean>(false);
+  //useRef
+  const [name, setName] = useState<string>("");
```
Comment (valentin-aleksandrov) [inline src/components/WorkBuddies/WorkBuddies.tsx]: Since you are providing an initial value, TS is smart enough to figure out the type on its own. This mean we can simplify:
```suggestion
  const [name, setName] = useState("");
```

File: src/components/WorkBuddies/WorkBuddies.tsx [hunk 27]
```diff
@@ -0,0 +1,150 @@
...
+import classNames from "classnames";
+import { Buddy } from "./BuddyInterface";
+import { BuddiesContext } from "../BuddiesContext";
+import MyWorkBuddy from "./MyWorkBuddy";
+
+const WorkBuddies = () => {
+  const [buddies, setBuddies] = useState<Buddy[]>([]);
+  const [myBuddy, setMyBuddy] = useState<Buddy>();
+  const [isBuddyChosen, setIsBuddyChosen] = useState<boolean>(false);
+  //useRef
```
Comment (valentin-aleksandrov) [inline src/components/WorkBuddies/WorkBuddies.tsx]: I'm not sure if you need this comment, since when you search for useRef, VS Code will find it.
In case you want to keep this comment, you may want to place above line 22 where you are using "useRef"
```suggestion
```

File: src/components/FormContext.tsx [hunk 28]
```diff
@@ -0,0 +1,30 @@
+import { createContext, useEffect, useState } from "react";
+
+interface FormDisabledContextType {
+  isDisabled: boolean | undefined;
+  setDisabled: () => void;
+}
+
+export const FormDisabledContext = createContext<FormDisabledContextType>({
+  isDisabled: false,
+  setDisabled: () => {},
```
Comment (valentin-aleksandrov) [inline src/components/FormContext.tsx]: In this scenario, I don't think this is the correct name for this function.
The function would toggle the state. Whoever is using the function wouldn't be allowed set the state for something specific.
I won't be able to do the following:
```jsx
setDisabled(true);
```
Do you think we can choose a more accurate name?
```suggestion
  toggleDisabled: () => {},
```

File: src/components/Tasks/tasks.css [hunk 29]
```diff
@@ -137,6 +143,16 @@ input {
   box-sizing: border-box;
   outline: #2c6830;
 }
+.buddy-smaller img{
```
Comment (valentin-aleksandrov) [inline src/components/Tasks/tasks.css]: Tea, could you enable your prettier on saving, so that you code can get formatted automatically?
```suggestion
.buddy-smaller img {
```

File: src/components/FormContext.tsx (hunk 28 above)
Comment (TheaAleksovDev) [reply src/components/FormContext.tsx]: Yes. Sounds good.

File: src/components/Tasks/tasks.css (hunk 29 above)
Comment (TheaAleksovDev) [reply src/components/Tasks/tasks.css]: Yes!

**Review by valentin-aleksandrov — APPROVED** [review]
